magda-semantic-indexer-framework/src/onRecordFoundStorageObject.ts

```diff
@@ -0,0 +1,164 @@
+                                {
+                                    record,
+                                    format: format,
+                                    filePath: filePath,
+                                    url: downloadURL
+                                }
+                            );
+                        } finally {
+                            if (filePath) {
+                                await deleteTempFile(filePath);
+                            }
+                        }
+                    } else {
+                        embeddingText = await userConfig.createEmbeddingText({
+                            record,
+                            format: format,
+                            filePath: null,
+                            url: downloadURL
+                        });
+                    }
+                    try {
+                        await indexEmbeddingText(
+                            userConfig,
+                            embeddingText,
+                            {
+                                recordId: record.id,
+                                fileFormat: format
+                            },
+                            chunker,
+                            embeddingApiClient,
+                            opensearchApiClient
+                        );
+                    } catch (err) {
+                        throw new SkipError(
+                            `Failed to index embedding text, error: ${
+                                (err as Error).message
+                            }`
+                        );
+                    }
+                } catch (err) {
+                    if (err instanceof SkipError) {
+                        console.warn("Skipping record because:", err.message);
+                        return;
+                    }
+                    throw err;
+                }
+            });
+            await Promise.all(tasks);
+        } catch (err) {
+            if (err instanceof SkipError) {
+                console.warn("Skipping record because:", err.message);
+                return;
+            }
+            throw err;
+        }
+    };
+};
+
+async function downloadFileWithRetry(url: string): Promise<string> {
+    return retry(
+        () => downloadFile(url),
+        1,
+        5,
+        (err, retries) => {
+            console.warn(
+                `Failed to download file, error: ${err.message}, retries: ${retries}`
+            );
+        }
+    );
+}
+
+async function downloadFile(url: string): Promise<string> {
+    const response = await fetch(url);
+    if (!response.ok) {
+        throw new SkipError(`${response.status} ${response.statusText}`);
+    }
+
+    const tempDir = tmpdir();
+    const tempFileName = `${uuidv4()}`;
+    const tempFilePath = join(tempDir, tempFileName);
+
+    const fileStream = await fs.open(tempFilePath, "w");
```

Comment:
performance wise, use fs.createWriteStream to create a writable stream and pipe in the response.body (readable stream) would be better.
e.g. something like: response.body.pipe(fs.createWriteStream(filepath))